File: LeetCode/986. Interval List Intersections.py
# https://leetcode.com/problems/interval-list-intersections/description/
from test import test
from typing import List


class Solution:
    def intervalIntersection(
        self, list1: List[List[int]], list2: List[List[int]]
    ) -> List[List[int]]:
        result: List[List[int]] = []

        index1 = 0
        index2 = 0

        while all([index1 < len(list1), index2 < len(list2)]):
            start1, finish1 = list1[index1]
            start2, finish2 = list2[index2]

            if finish1 < start2:
                index1 += 1
                continue
            if finish2 < start1:
                index2 += 1
                continue

            pair = [max(start1, start2), min(finish1, finish2)]

            result.append(pair)

            if finish1 > finish2:
                index2 += 1
            else:
                index1 += 1

        return result


solution = Solution()


test(
    solution.intervalIntersection,
    [
        {
            "input": [
                [
                    [0, 2],
                    [5, 10],
                    [13, 23],
                    [24, 25],
                ],
                [
                    [1, 5],
                    [8, 12],
                    [15, 24],
                    [25, 26],
                ],
            ],
            "expected": [
                [1, 2],
                [5, 5],
                [8, 10],
                [15, 23],
                [24, 24],
                [25, 25],
            ],
        },
        {
            "input": [
                [
                    [1, 3],
                    [5, 9],
                ],
                [],
            ],
            "expected": [],
        },
    ],
)

File: LeetCode/986. Interval List Intersections.js
// https://leetcode.com/problems/interval-list-intersections/description/

const { test } = require('../../test');

const intervalIntersection = function (arr1, arr2) {
  const result = [];

  let index1 = 0;
  let index2 = 0;

  while (index1 < arr1.length && index2 < arr2.length) {
    const [start1, finish1] = arr1[index1];
    const [start2, finish2] = arr2[index2];

    if (start2 > finish1) {
      index1++;
      continue;
    }

    if (start1 > finish2) {
      index2++;
      continue;
    }

    const pair = [Math.max(start1, start2), Math.min(finish1, finish2)];

    result.push(pair);

    if (finish1 > finish2) {
      index2++;
    } else {
      index1++;
    }
  }

  return result;
};

test(intervalIntersection, [
  {
    input: [
      [
        [0, 2],
        [5, 10],
        [13, 23],
        [24, 25],
      ],
      [
        [1, 5],
        [8, 12],
        [15, 24],
        [25, 26],
      ],
    ],
    expected: String([
      [1, 2],
      [5, 5],
      [8, 10],
      [15, 23],
      [24, 24],
      [25, 25],
    ]),
  },
  {
    input: [
      [
        [1, 3],
        [5, 9],
      ],
      [],
    ],
    expected: '',
  },
]);
